File: bindings/pybind11_ext/test/test_numpy_dtypes_user.cpp
/*
  tests/test_numpy_dtypes.cpp -- User defined NumPy dtypes

  Copyright (c) 2018 Eric Cousineau

  All rights reserved. Use of this source code is governed by a
  BSD-style license that can be found in the LICENSE file.
*/

#include <cstring>

#include <pybind11/stl.h>
#include <pybind11/operators.h>

#include "pybind11_tests.h"
#include "constructor_stats.h"

#include "pybind11_ext/numpy_dtypes_user.h"

using std::string;
using std::unique_ptr;

namespace py = pybind11;

namespace {

// Trivial string class.
class CustomStr {
public:
    static constexpr int len = 100;
    CustomStr(const char* s) {
        snprintf(buffer.data(), len, "%s", s);
        dummy.reset(new int(1000));
    }
    template <typename Arg, typename... Args>
    CustomStr(const char* fmt, Arg arg, Args... args) {
        snprintf(buffer.data(), len, fmt, arg, args...);
        dummy.reset(new int(1000));
    }
    CustomStr(const CustomStr&) = default;
    CustomStr& operator=(const CustomStr&) = default;
    CustomStr(CustomStr&&) = default;
    CustomStr& operator=(CustomStr&&) = default;
    std::string str() const {
        return buffer.data();
    }
    bool operator==(const CustomStr& other) const {
        return str() == other.str();
    }
private:
    std::array<char, len> buffer;
    // Data member to ensure that we do not get segfaults when carrying around `shared_ptr`s.
    std::shared_ptr<int> dummy;
};

// Basic structure, meant to be an implicitly convertible value for `Custom`.
// Would have used a struct type, but the scalars are only tuples.
struct SimpleStruct {
    double value;

    SimpleStruct(double value_in) : value(value_in) {}
};

// Define operations to return fixed values.
double operator+(double, SimpleStruct) {
    return -1;
}
double operator+(SimpleStruct, SimpleStruct) {
    return 0;
}
double operator+(SimpleStruct, double) {
    return 1;
}

// Clones a unique_ptr using type's copy constructor.
template <typename T>
void clone(const unique_ptr<T>& src, unique_ptr<T>& dst) {
    if (!src)
        dst.reset();
    else
        dst.reset(new T(*src));
}

class Custom {
public:
    Custom() {
        print_created(this);
    }
    ~Custom() {
        track_destroyed(this);
        // TODO(eric.cousineau): Track down why this causes a segfault.
        //print_destroyed(this);
    }
    Custom(double value) : value_{value} {
        print_created(this, value);
    }
    Custom(double value, string str)
        : value_{value}, str_{unique_ptr<string>(new string(str))} {
        print_created(this, value, str);
    }
    Custom(const Custom& other) {
        value_ = other.value_;
        clone(other.str_, str_);
        print_copy_created(this, other.value_);
    }
    Custom(const SimpleStruct& other) {
        value_ = other.value;
        print_copy_created(this, other);
    }
    Custom& operator=(const Custom& other) {
        py::gil_scoped_acquire acq;
        print_copy_assigned(this, other.value_);
        value_ = other.value_;
        clone(other.str_, str_);
        return *this;
    }

    double value() const { return value_; }

    operator SimpleStruct() const { return {value_}; }

    Custom operator+(const Custom& rhs) const {
        auto tmp = Custom(value_ + rhs.value_);
        return tmp;
    }
    Custom& operator+=(const Custom& rhs) {
        value_ += rhs.value_;
        return *this;
    }
    Custom operator+(double rhs) const {
        auto tmp = Custom(value_ + rhs);
        return tmp;
    }
    Custom& operator+=(double rhs) {
        value_ += rhs;
        return *this;
    }
    Custom operator*(const Custom& rhs) const { return value_ * rhs.value_; }
    Custom operator-(const Custom& rhs) const { return value_ - rhs.value_; }

    Custom operator-() const { return -value_; }

    Custom operator/(const Custom& rhs) const {
        return Custom(value_ / rhs.value_);
    }

    bool same_as(const Custom& rhs) const {
        return value_ == rhs.value_ && str() == rhs.str();
    }
    CustomStr operator==(const Custom& rhs) const {
        // Return non-boolean dtype.
        return CustomStr("%g == %g && '%s' == '%s'", value_, rhs.value_, str().c_str(), rhs.str().c_str());
    }
    bool operator<(const Custom& rhs) const {
        // Return boolean value.
        return value_ < rhs.value_;
    }

    std::string str() const {
        if (str_)
            return *str_;
        else
            return {};
    }

private:
    double value_{};
    // Use non-trivial data object, but something that is memcpy-movable.
    std::unique_ptr<string> str_;
};

Custom operator+(double a, const Custom& b) {
    return Custom(a) += b;
}

// Define `dtype=object` types, and ensure the operations correctly reflect these.
struct ObjectA {};
struct ObjectB {};

Custom operator+(const Custom&, const ObjectA&) {
    return Custom(1000);
}
Custom operator+(const Custom&, const ObjectB&) {
    return Custom(9999);
}

class Container {
public:
    Container() {
        value_ = {{Custom(0), Custom(1)}};
    }
    std::array<Custom, 2>& value() { return value_; }
private:
    std::array<Custom, 2> value_;
};

class ImplicitArg {
public:
    ImplicitArg() {}
    ImplicitArg(double value) : value_(value) {}
    double value() const { return value_; }
private:
    double value_{};
};

}  // namespace

PYBIND11_NUMPY_DTYPE_USER(CustomStr);
PYBIND11_NUMPY_DTYPE_USER(SimpleStruct);
PYBIND11_NUMPY_DTYPE_USER(Custom);
PYBIND11_NUMPY_DTYPE_USER(ImplicitArg);

PYBIND11_NUMPY_OBJECT_DTYPE(ObjectA);
PYBIND11_NUMPY_OBJECT_DTYPE(ObjectB);

TEST_SUBMODULE(numpy_dtype_user, m) {
    // Supress warnings
    (void)py::detail::npy_format_descriptor<ObjectA>::name;
    (void)py::detail::npy_format_descriptor<ObjectB>::name;

    ConstructorStats::type_fallback([](py::object cls) -> std::type_index* {
        return const_cast<std::type_index*>(py::detail::dtype_info::find_entry(cls));
    });

    try { py::module::import("numpy"); }
    catch (...) { return; }

    py::class_<ObjectA>(m, "ObjectA")
        .def(py::init());
    py::class_<ObjectB>(m, "ObjectB")
        .def(py::init());

    // Bare, minimal type.
    py::dtype_user<CustomStr>(m, "CustomStr")
        .def(py::init<const char*>())
        .def("__str__", &CustomStr::str)
        .def("__repr__", &CustomStr::str)
        .def_loop(py::dtype_method::explicit_conversion(
            [](const CustomStr&) -> double {
                py::pybind11_fail("Cannot cast");
            }))
        .def(py::self == py::self);

    // Not explicitly convertible: `double`
    auto ss_str = [](const SimpleStruct& self) {
        return py::str("SimpleStruct({})").format(self.value);
    };
    py::dtype_user<SimpleStruct> simple_struct(m, "SimpleStruct");
    py::dtype_user<Custom> custom(m, "Custom");
    simple_struct
        .def(py::init<double>())
        .def("__str__", ss_str)
        .def("__repr__", ss_str)
        // Test operator ordering.
        .def_loop(py::self + py::self)
        .def_loop(double{} + py::self)
        .def_loop(py::self + double{})
        .def_loop(
            py::dtype_method::implicit_conversion<SimpleStruct, Custom>());

    custom
        .def(py::init())
        .def(py::init<double>())
        .def(py::init<double, string>())
        .def(py::init<const SimpleStruct&>())
        .def(py::init<Custom>())
        .def("__repr__", [](const Custom* self) {
            return py::str("Custom({}, '{}')").format(self->value(), self->str());
        })
        .def("__str__", [](const Custom* self) {
            return py::str("C<{}, '{}'>").format(self->value(), self->str());
        })
        .def("value", &Custom::value)
        // Test referencing.
        .def("self", [](Custom* self) { return self; }, py::return_value_policy::reference)
        // Casting.
        // N.B. For `np.ones`, we could register a converter from `int64_t` to `Custom`, but this would cause a segfault,
        // because `np.ones` uses `np.copyto(..., casting="unsafe")`, which does *not* respect NPY_NEEDS_INITIALIZATION.
        // - Explicit casting (e.g., we have additional arguments).
        .def_loop(py::dtype_method::explicit_conversion(
            [](const Custom& in) -> double { return in.value(); }))
        .def_loop(py::dtype_method::explicit_conversion<double, Custom>())
            // - Implicit coercion + conversion
        .def_loop(py::dtype_method::implicit_conversion(
            &Custom::operator SimpleStruct))
           // - - N.B. This shouldn't be a normal operation (upcasting?), as it may result in data loss.
        .def_loop(py::dtype_method::implicit_conversion<SimpleStruct, Custom>())
            // TODO(eric.cousineau): Figure out type for implicit coercion.
            // Operators + ufuncs, with some just-operators (e.g. in-place)
        .def_loop(py::self + py::self)
        .def(py::self += py::self)
        .def_loop(py::self + double{})
        .def_loop(double{} + py::self)
        .def_loop(py::self / py::self)
        .def(py::self += double{})
        .def_loop(py::self * py::self)
        .def_loop(py::self - py::self)
        .def_loop(-py::self)
        .def_loop(py::self == py::self)
        .def_loop(py::self < py::self)
        .def(py::self + ObjectA{})
        .def(py::self + ObjectB{})
        .def_loop(py::dtype_method::dot())
        .def_loop("__pow__", [](const Custom& a, const Custom& b) {
            return CustomStr("%g ^ %g", a.value(), b.value());
        })
        // TOOD(eric.cousineau): Handle pointers too.
        .def_loop("cos", [](const Custom& self) {
            return Custom(cos(self.value()));
        })
        .def_loop("logical_and", [](const Custom&, const Custom&) -> double {
            return 10;
        })
        .def("str", &Custom::str);
    // Somewhat more expressive.

    auto binary_op = [](const Custom& a, const Custom& b) {
        return CustomStr("%g == %g", a.value(), b.value());
    };
    m.def("binary_op", binary_op);
    py::ufunc(m, "binary_op_loop")
        .def_loop<Custom>(binary_op);

    // N.B. We should not define a boolean operator for `equal`, as NumPy will
    // use this, even if we define it "afterwards", due to how it is stored.

    // `py::vectorize` does not seem to allow custom types due to sfinae constraints :(
    py::ufunc x(m, "same");
    x
        .def_loop<Custom>([](const Custom& a, const Custom& b) {
            return a.same_as(b);
        })
        .def_loop<CustomStr>([](const CustomStr& a, const CustomStr& b) {
            return a == b;
        })
        // Define this for checking other stuff.
        .def_loop<bool>([](bool a, bool b) { return a == b; })
        .def_loop<double>([](double a, double b) { return a == b; })
        .def_loop<py::object>([m](py::object a, py::object b) {
            // Will recurse if we don't have an overload for the given type.
            return m.attr("same")(a, b).cast<bool>();
        });

    py::class_<Container>(m, "Container")
        .def(py::init())
        .def("value", [](py::handle self) {
            auto& me = self.cast<Container&>();
            return py::array_t<Custom>({1, 2}, me.value().data(), self);
        }, py::return_value_policy::reference_internal);

    m.def("add_one", [](py::handle in) {
        auto value = py::cast<py::array_t<Custom>>(in.attr("flat"));
        int count = static_cast<int>(value.size());
        for (int i = 0; i < count; ++i) {
            auto& x = value.mutable_at(i);
            x += 1;
        }
    });

    py::module np = py::module::import("numpy");
    py::dtype np_int64 = py::reinterpret_borrow<py::dtype>(
        np.attr("dtype")(np.attr("int64")));
    py::dtype np_float16 = py::reinterpret_borrow<py::dtype>(
        np.attr("dtype")(np.attr("float16")));

    py::dtype_user<ImplicitArg>(m, "ImplicitArg")
        .def(py::init())
        .def(py::init<double>())
        .def_loop(
            py::dtype_method::implicit_conversion([](int64_t in) -> ImplicitArg {
                double value = static_cast<double>(in);
                return value;
            }), np_int64)
        .def_loop(
            py::dtype_method::implicit_conversion<double, ImplicitArg>())
        .def_loop(
            py::dtype_method::implicit_conversion([](int16_t) -> ImplicitArg {
                throw std::runtime_error("No actual half float conversion available");
            }), np_float16);

    m.def("implicit_arg_scalar", [](ImplicitArg in) { return in; });
    m.def("implicit_arg_vector", [](py::array_t<ImplicitArg> in) { return in; });
}

File: bindings/pybind11_ext/test/test_numpy_dtypes_user.py
import gc
import os

import numpy as np
import pytest

from pybind11_tests import ConstructorStats
import pybind11_tests.numpy_dtype_user as m


def test_scalar_meta():
    """Tests basic metadata."""
    assert issubclass(m.Custom, np.generic)
    assert isinstance(np.dtype(m.Custom), np.dtype)


def test_scalar_ctor():
    """Tests instance lifetime management since we had to redo the instance
    registry to inherit from `np.generic` :( """
    c = m.Custom()
    c1 = m.Custom(c)
    assert id(c) != id(c1)
    assert id(c.self()) == id(c)
    del c
    del c1
    gc.collect()
    stats = ConstructorStats.get(m.Custom)
    assert stats.alive() == 0


def test_scalar_functions():
    # Test scalar functions.
    a = m.Custom(1)
    assert repr(a) == "Custom(1.0, '')"
    assert str(a) == "C<1.0, ''>"
    assert m.same(a, a)
    b = m.Custom(2)
    assert not m.same(a, b)


def test_scalar_algebra():
    """Tests scalar algebra."""
    a = m.Custom(1)
    b = m.Custom(2)
    # N.B. Implicit casting is not easily testable here; see array tests.
    # Operators.
    # - self + self
    assert m.same(a + b, m.Custom(3))
    a += b
    assert m.same(a, m.Custom(3))
    # - self + double (and int, implicitly)
    assert m.same(a + 2, m.Custom(5))
    assert m.same(2 + a, m.Custom(5))
    a += 2.
    a += 1
    assert m.same(a, m.Custom(6))
    a = m.Custom(6)
    # Others.
    assert m.same(a * b, m.Custom(12))
    assert m.same(a - b, m.Custom(4))
    assert m.same(-a, m.Custom(-6))
    # Logical.
    assert m.same(a == b, m.CustomStr("6 == 2 && '' == ''"))
    assert m.same(a < b, False)


def test_array_creation():
    # Zeros.
    x = np.zeros((2, 2), dtype=m.Custom)
    assert x.shape == (2, 2)
    assert x.dtype == m.Custom
    # Generic creation.
    x = np.array([m.Custom(1, "Howdy")])
    assert x.dtype == m.Custom
    # - Limitation on downcasting when mixing types.
    # This could be alleviated by allowing doubles to be implicitly casted for
    # the type, but it's best to avoid that.
    x = np.array([m.Custom(10), 1.])
    assert x.dtype == object
    # - At present, we will be leaking memory. This doesn't show up in instance
    # count, since these items are only mutated via `operator=`; however, it will
    # still be the case for resizing.
    # See https://github.com/numpy/numpy/issues/10721 for more information.


def test_array_creation_extended():
    with pytest.raises(ValueError):
        # Fails due to `np.copyto` relying on `np.long` conversion on uninitialized memory.
        x = np.ones((2, 2), dtype=m.Custom)
    x = np.ones((1, 2)).astype(m.Custom)
    assert check_array(x, [[m.Custom(1), m.Custom(1)]])
    x = np.full((1, 2), m.Custom(10), dtype=m.Custom)
    assert check_array(x, [[m.Custom(10), m.Custom(10)]])
    # `np.eye(..., dtype=m.Custom)` requires a converter from `int` to
    # `m.Custom` (or something that tells it to use `double`).
    # Prefer to avoid, as it complicates other implicit conversions,
    # which in general shouldn't be there, but nonetheless should be tested.
    x = np.eye(2).astype(m.Custom)
    assert check_array(x, [[m.Custom(1), m.Custom(0)], [m.Custom(0), m.Custom(1)]])


def check_array(actual, expected):
    """Checks if two arrays are exactly similar (shape, type, and data)."""
    expected = np.array(expected)
    if actual.shape != expected.shape:
        return False
    if actual.dtype != expected.dtype:
        return False
    if not m.same(actual, expected).all():
        return False
    return True


def test_array_cast():
    def check(x, dtype):
        dx = x.astype(dtype)
        assert dx.dtype == dtype, dtype
        assert dx.astype(m.Custom).dtype == m.Custom
        return dx
    # Custom -> {Custom, float, object}
    x = np.array([m.Custom(1)])
    check(x, m.Custom)
    check(x, float)
    check(x, object)
    # float -> Custom
    x = np.array([1., 2.])
    check(x, m.Custom)
    # object -> Custom
    # - See notes in the C++ code for defining the ufunc cast for `object` to
    # `Class`.
    x = np.array([1., m.Custom(10), m.SimpleStruct(100)], dtype=object)
    dx = check(x, m.Custom)
    assert check_array(dx, [m.Custom(1), m.Custom(10), m.Custom(100)])


def test_array_cast_implicit():
    a = np.array([1.]).astype(m.Custom)
    # - We registered `Custom{} + double{}`.
    a += 2
    assert check_array(a, [m.Custom(3.)])
    # - Try multiple elements.
    a = np.array([1., 2.]).astype(m.Custom)
    a += 2.
    assert check_array(a, [m.Custom(3.), m.Custom(4.)])
    # We do not allow implicit coercion from `double` to Custom:
    with pytest.raises(TypeError):
        a[0] = 1.
    with pytest.raises(TypeError):
        b = np.array([1., 2.], dtype=m.Custom)
        print(b)  # Supress "unused" warnings
    with pytest.raises(TypeError):
        a *= 2
    # Try coercion from `Custom` to `double`
    af = np.array([0., 0.], dtype=float)
    # The following is viewed as "explicit" since we're using a slice, I guess?
    af[:] = a
    assert check_array(af, [3., 4.])
    af[:] = af + a
    assert check_array(af, [6., 8.])
    # This, I guess, tries explicit coercion, which is not available.
    with pytest.raises(TypeError):
        af += a
    # Also causes an error, as it attempts implicit coercion, which we've
    # disabled.
    with pytest.raises(TypeError):
        af[0] = a[0]
    # - Try an unregistered type, e.g. `int`.
    ai = np.array([0, 0], dtype=int)
    with pytest.raises(ValueError):
        ai[:] = a
    with pytest.raises(TypeError):
        ai += a
    with pytest.raises(TypeError):
        ai[0] = a[0]
    # Try an implicit conversion.
    a = np.array([1.]).astype(m.Custom)
    # - Nominal pybind implicit conversion
    a[0] = m.SimpleStruct(9)
    assert check_array(a, [m.Custom(9)])
    # - Test array construction (numpy coercion)
    c = np.array([m.SimpleStruct(10), m.SimpleStruct(11)], dtype=m.Custom)
    assert check_array(c, [m.Custom(10), m.Custom(11)])
    # - Test implicit cast via coercion.
    c *= m.SimpleStruct(2)
    assert check_array(c, [m.Custom(20), m.Custom(22)])
    # - Show dangers of implicit conversion (info loss).
    ds = m.Custom(100, "Hello")
    d = np.array([ds])
    e = np.array([m.SimpleStruct(0)])
    e[:] = d
    d[:] = e
    assert not check_array(d, [ds])
    assert check_array(d, [m.Custom(100)])


def test_array_ufunc():
    x = np.array([m.Custom(4)])
    y = np.array([m.Custom(2, "World")])
    assert check_array(x + y, [m.Custom(6)])
    assert check_array(x + 1, [m.Custom(5)])
    assert check_array(1 + x, [m.Custom(5)])
    assert check_array(x * y, [m.Custom(8)])
    assert check_array(x - y, [m.Custom(2)])
    assert check_array(x / y, [m.Custom(2)])
    assert check_array(-x, [m.Custom(-4)])
    assert check_array(x == y, [m.CustomStr("4 == 2 && '' == 'World'")])
    assert check_array(x < y, [False])
    assert check_array(np.power(x, y), [m.CustomStr("4 ^ 2")])
    assert check_array(np.dot(x, y), m.Custom(8))
    assert check_array(np.dot([x], [y]), [[m.Custom(8)]])
    assert check_array(np.cos(x), [m.Custom(np.cos(x[0].value()))])
    assert check_array(np.logical_and(x, y), [10.])


def test_array_op_order():
    s = m.SimpleStruct(1)
    f = 0.
    assert f + s == -1
    assert s + s == 0
    assert s + f == 1
    sv = np.array([s])
    fv = np.array([f])
    assert check_array(sv + sv, [0.])
    assert check_array(sv + fv, [1.])
    assert check_array(fv + sv, [-1.])


def test_object_mixing():
    # Ensure that operations involving `dtype=object` downcasts properly.
    c = m.Custom(0)
    a = m.ObjectA()
    b = m.ObjectB()
    cv = np.array([c])
    av = np.array([a])
    bv = np.array([b])
    # These operations are now downcast to `object`.
    assert check_array(cv + av, np.array([m.Custom(1000.)], dtype=object))
    assert check_array(cv + bv, np.array([m.Custom(9999.)], dtype=object))


def test_implicit_arguments():
    s1 = m.SimpleStruct(1)
    s2 = m.SimpleStruct(1000)
    s1a = np.array(s1)
    s2a = np.array(s2)
    y = m.binary_op(s1, s2)
    assert m.same(y, m.CustomStr("1 == 1000"))
    y = m.binary_op(s1a, s2a)
    assert m.same(y, m.CustomStr("1 == 1000"))
    with pytest.raises(TypeError):
        # This does not work, even when declaring implicit arguments. Most likely
        # because NumPy needs to know an anchoring type?
        y = m.binary_op_loop(s1a, s2a)
    # The following works because NumPy is aware of the type...
    c1 = m.Custom(s1)
    y = m.binary_op_loop(c1, s2a)
    assert m.same(y, m.CustomStr("1 == 1000"))


def test_reference_arguments():
    # Test Python -> C++ reference
    x = np.array([m.Custom(0), m.Custom(1)])
    m.add_one(x)
    assert check_array(x, [m.Custom(1), m.Custom(2)])
    # Test C++ -> Python reference
    c = m.Container()
    xc = c.value()
    assert xc.shape == (1, 2)
    xc += 10
    assert check_array(c.value(), [
        [m.Custom(10), m.Custom(11)]])
    m.add_one(c.value())
    assert check_array(c.value(), [
        [m.Custom(11), m.Custom(12)]])


def test_copy():
    x = np.array([m.Custom(1, "a")])
    y = np.copy(x)
    x[0] = m.Custom(10, "c")
    assert x[0].value() == 10
    assert x[0].str() == "c"
    assert y[0].value() == 1
    assert y[0].str() == "a"


def test_implicit_arg():
    # Test implicit casting from an integer.
    x = m.implicit_arg_scalar(1)
    assert isinstance(x, m.ImplicitArg)
    iv = np.array([1, 2])
    assert iv.dtype == np.int64
    xv = m.implicit_arg_vector(iv)
    assert xv.dtype == m.ImplicitArg
    assert xv.shape == (2,)
    m.implicit_arg_vector([1, 2])
    m.implicit_arg_vector([1., 2.])
    m.implicit_arg_vector(np.array([1., 2.]))


def test_result_type():
    dt = np.result_type(m.ImplicitArg, np.int64)
    assert dt == m.ImplicitArg
    dt = np.result_type(m.ImplicitArg, np.float)
    assert dt == m.ImplicitArg
    dt = np.result_type(m.ImplicitArg, 1.)
    assert dt == m.ImplicitArg


if __name__ == "__main__":
    pytest.main()
